File: JTSArjuna/JTSArjuna-1.2.4/JavaGandiva/classes/com/arjuna/JavaGandiva/Template/BasicListIterator.java
/*
 * Copyright (C) 1996, 1997, 1998,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: BasicListIterator.java,v 1.2 1998/07/06 13:31:03 nmcl Exp $
 */

package com.arjuna.JavaGandiva.Template;

public class BasicListIterator
{

public BasicListIterator (BasicList l)
    {
	_list = l;
	_current = _list._head;
    }

public synchronized Object next ()
    {
	if (_current != null)
	{
	    Object ptr = _current.value();

	    _current = _current.next();

	    return ptr;
	}
	else
	    return null;
    }

private BasicList _list;
private ObjectElement _current;
    
};

File: JTSArjuna/JTSArjuna-1.2.4/Gandiva/include/Template/BasicListIterator.cc
/*
 * Copyright (C) 1994, 1995, 1996, 1997,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: BasicListIterator.cc,v 1.5 1997/07/01 14:55:40 nsmw Exp $
 */

#ifndef TEMPLATE_BASICLISTITERATOR_CC_
#define TEMPLATE_BASICLISTITERATOR_CC_

template <class T, class U>
BasicListIterator<T,U>::BasicListIterator (const BasicList<T, U>& L)
					: ptr(L.Head)
{
}

template <class T, class U>
T* BasicListIterator<T,U>::operator ()()
{
    if (ptr)
    {
	ListCons<T>* p = ptr;
	ptr = ptr->cdr();
	
	return p->car();
    }

    return (T*) 0;
}

#endif
